bunsen/bunsen-core/src/main/java/com/cerner/bunsen/ProfileMapperFhirContexts.java

```diff
@@ -0,0 +1,179 @@
+ */
+public class ProfileMapperFhirContexts {
+
+  private static final Logger logger =
+      LoggerFactory.getLogger(ProfileMapperFhirContexts.class.getName());
+  private final Map<FhirVersionEnum, FhirContextData> fhirContextMappings;
+  private static ProfileMapperFhirContexts instance;
+  private ProfileMappingProvider profileMappingProvider;
+
+  private ProfileMapperFhirContexts() {
+    this.fhirContextMappings = Maps.newHashMap();
+    this.profileMappingProvider = new ProfileMappingProvider();
+  }
+
+  /** This method returns a singleton instance of the current class */
+  public static ProfileMapperFhirContexts getInstance() {
+    synchronized (ProfileMapperFhirContexts.class) {
+      if (instance == null) {
+        instance = new ProfileMapperFhirContexts();
+      }
+      return instance;
+    }
+  }
+
+  /**
+   * Returns the FHIR context for the given version and the structureDefinitionsPath. For the given
+   * fhirVersion, this method loads all the base profiles and also the custom profiles defined in
+   * the structureDefinitionsPath. This method caches the Fhir context for the given input
+   * combination of fhirVersion and structureDefinitionsPath, and returns the same FhirContext for
+   * any further invocations of similar arguments. It returns an error if the method is called with
+   * a different structureDefinitionsPath for the same fhirVersion earlier.
+   *
+   * @param fhirVersion the version of FHIR to use
+   * @param structureDefinitionsPath The path containing the custom structure definitions
+   * @return the FhirContext
+   * @throws ProfileMapperException if there are any errors while loading and mapping the structure
+   *     definitions
+   */
+  public synchronized FhirContext contextFor(
+      FhirVersionEnum fhirVersion, @Nullable String structureDefinitionsPath)
+      throws ProfileMapperException {
+    return contextFor(fhirVersion, structureDefinitionsPath, false);
+  }
+
+  /**
+   * Similar to {@link #contextFor(FhirVersionEnum, String)} but the custom profiles are loaded from
+   * the classpath name passed via structureDefinitionsClasspath.
+   */
+  public synchronized FhirContext contextForFromClasspath(
+      FhirVersionEnum fhirVersion, @Nullable String structureDefinitionsClasspath)
+      throws ProfileMapperException {
+    return contextFor(fhirVersion, structureDefinitionsClasspath, true);
+  }
+
+  private FhirContext contextFor(
+      FhirVersionEnum fhirVersion, @Nullable String structureDefinitionsPath, boolean isClasspath)
+      throws ProfileMapperException {
+    structureDefinitionsPath = Strings.nullToEmpty(structureDefinitionsPath);
+    FhirContextData fhirContextData = fhirContextMappings.get(fhirVersion);
+    if (fhirContextData != null) {
+      isContextLoadedWithDifferentConfig(fhirContextData, structureDefinitionsPath, isClasspath);
+    } else {
+      FhirContext context = new FhirContext(fhirVersion);
```

Comment:
I understand we have guards against recreating `FhirContext` objects, but still should we use `FhirContext.forCached()` method (in HAPI) instead? I mean are there any drawbacks?

Reply:
I don't see a drawback, we can use this function to create a single FhirContext for the given version. However, I am not changing the current code, because I need to load different StructureDefinitions for different Test and by static context I am not able to load the second ones.

Reply:
Okay then if this is required, please feel free to resolve this comment (maybe also add a single line comment before this why we are not using the cached version).